### PR TITLE
http server and client builder for netty

### PR DESCRIPTION
**Pull request by zeitlinger**

### CODE REVIEW

**Comment by zeitlinger** [comment]
@trask @laurit can you take a look?

File: instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpClientInstrumenterBuilder.java [hunk 1]
```diff
@@ -222,6 +222,10 @@ public OpenTelemetry getOpenTelemetry() {
     return openTelemetry;
   }
 
+  public String getInstrumentationName() {
+    return instrumentationName;
+  }
```
Comment (laurit) [inline instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpClientInstrumenterBuilder.java:227]: to me this method and `getOpenTelemetry` feel out of place in this class.

File: instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpServerInstrumenterBuilder.java [hunk 2]
```diff
@@ -177,7 +185,7 @@ public Instrumenter<REQUEST, RESPONSE> build() {
     return builder.buildInstrumenter(SpanKindExtractor.alwaysServer());
   }
 
-  private InstrumenterBuilder<REQUEST, RESPONSE> builder() {
+  public InstrumenterBuilder<REQUEST, RESPONSE> builder() {
```
Comment (laurit) [inline instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpServerInstrumenterBuilder.java:188]: I could not find any usages for this method outside of this class.

File: instrumentation/netty/netty-4-common/library/src/main/java/io/opentelemetry/instrumentation/netty/v4/common/internal/client/NettyClientInstrumenterBuilderFactory.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import io.netty.handler.codec.http.HttpResponse;
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.instrumentation.api.incubator.builder.internal.DefaultHttpClientInstrumenterBuilder;
+import io.opentelemetry.instrumentation.netty.v4.common.HttpRequestAndChannel;
+
+/**
+ * This class is internal and is hence not for public use. Its APIs are unstable and can change at
+ * any time.
+ */
+public class NettyClientInstrumenterBuilderFactory {
```
Comment (laurit) [inline instrumentation/netty/netty-4-common/library/src/main/java/io/opentelemetry/instrumentation/netty/v4/common/internal/client/NettyClientInstrumenterBuilderFactory.java:17]: by convention I think this class should also be final

File: instrumentation/netty/netty-4.1/library/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/NettyServerTelemetryBuilder.java [hunk 4]
```diff
@@ -8,32 +8,32 @@
...
-      extractorConfigurer = builder -> {};
-  private Consumer<HttpSpanNameExtractorBuilder<HttpRequestAndChannel>>
-      spanNameExtractorConfigurer = builder -> {};
-  private Consumer<HttpServerRouteBuilder<HttpRequestAndChannel>> httpServerRouteConfigurer =
-      builder -> {};
-  private boolean emitExperimentalHttpServerMetrics = false;
   private boolean emitExperimentalHttpServerEvents = false;
 
+  static {
+    NettyServerInstrumenterBuilderUtil.setBuilderExtractor(NettyServerTelemetryBuilder::getBuilder);
```
Comment (laurit) [inline instrumentation/netty/netty-4.1/library/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/NettyServerTelemetryBuilder.java:29]: could use `NettyServerInstrumenterBuilderUtil.setBuilderExtractor(nettyServerTelemetryBuilder -> nettyServerTelemetryBuilder.builder);` so you wouldn't need to add the `getBuilder` method

File: instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpServerInstrumenterBuilder.java (hunk 2 above)
Comment (zeitlinger) [reply instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpServerInstrumenterBuilder.java:188]: added a `instrumenterBuilder` method instead and only exposed propagators of `OpenTelemetry`

File: instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpClientInstrumenterBuilder.java (hunk 1 above)
Comment (zeitlinger) [reply instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpClientInstrumenterBuilder.java:227]: added a `instrumenterBuilder` method instead and only exposed propagators of `OpenTelemetry`

File: instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpServerInstrumenterBuilder.java (hunk 2 above)
Comment (zeitlinger) [reply instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpServerInstrumenterBuilder.java:188]: inlined

**Comment by zeitlinger** [comment]
@laurit please check again

File: instrumentation/netty/netty-4.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/NettyServerSingletons.java [hunk 5]
```diff
@@ -7,21 +7,22 @@
...
-            .setCapturedRequestHeaders(AgentCommonConfig.get().getServerRequestHeaders())
-            .setCapturedResponseHeaders(AgentCommonConfig.get().getServerResponseHeaders())
-            .build();
+    NettyServerTelemetryBuilder builder = NettyServerTelemetry.builder(GlobalOpenTelemetry.get());
+    NettyServerInstrumenterBuilderUtil.getBuilderExtractor()
+        .apply(builder)
+        .configure(AgentCommonConfig.get());
+    if (AgentCommonConfig.get().shouldEmitExperimentalHttpServerTelemetry()) {
+      // this logic is only used in agent
+      builder.setEmitExperimentalHttpServerEvents(true);
```
Comment (trask) [inline instrumentation/netty/netty-4.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/NettyServerSingletons.java:23]: is there reason not to expose this to library instrumentation?

File: javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/bootstrap/internal/JavaagentHttpServerInstrumenters.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+      TextMapGetter<REQUEST> headerGetter,
+      Consumer<InstrumenterBuilder<REQUEST, RESPONSE>> instrumenterBuilderConsumer) {
+    return create(
+        new DefaultHttpServerInstrumenterBuilder<>(
+                instrumentationName, GlobalOpenTelemetry.get(), httpAttributesGetter)
+            .setHeaderGetter(headerGetter),
+        instrumenterBuilderConsumer);
+  }
+
+  public static <REQUEST, RESPONSE> Instrumenter<REQUEST, RESPONSE> create(
```
Comment (trask) [inline javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/bootstrap/internal/JavaagentHttpServerInstrumenters.java:48]:
```suggestion
  private static <REQUEST, RESPONSE> Instrumenter<REQUEST, RESPONSE> create(
```

File: instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpClientInstrumenterBuilder.java [hunk 7]
```diff
@@ -218,8 +219,14 @@ public Instrumenter<REQUEST, RESPONSE> build() {
...
-  public OpenTelemetry getOpenTelemetry() {
-    return openTelemetry;
+  public ContextPropagators getPropagators() {
+    return openTelemetry.getPropagators();
+  }
+
+  public <BUILDERREQUEST, BUILDERRESPONSE>
+      InstrumenterBuilder<BUILDERREQUEST, BUILDERRESPONSE> instrumenterBuilder(
+          SpanNameExtractor<? super BUILDERREQUEST> spanNameExtractor) {
+    return Instrumenter.builder(openTelemetry, instrumentationName, spanNameExtractor);
```
Comment (trask) [inline instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpClientInstrumenterBuilder.java:229]: this method is confusing, I finally understood it, but wonder if there's anything we can do to make it clearer

File: instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpServerInstrumenterBuilder.java [hunk 8]
```diff
@@ -195,8 +194,12 @@ private InstrumenterBuilder<REQUEST, RESPONSE> builder() {
           .addAttributesExtractor(HttpExperimentalAttributesExtractor.create(attributesGetter))
           .addOperationMetrics(HttpServerExperimentalMetrics.get());
     }
+    builderCustomizer.accept(builder);
 
-    return builder;
+    if (headerGetter != null) {
+      return builder.buildServerInstrumenter(headerGetter);
+    }
+    return builder.buildInstrumenter(SpanKindExtractor.alwaysServer());
```
Comment (trask) [inline instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpServerInstrumenterBuilder.java:202]: are there any server instrumentations that don't provide a `headerGetter`?

File: instrumentation/netty/netty-4.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/NettyServerSingletons.java (hunk 5 above)
Comment (zeitlinger) [reply instrumentation/netty/netty-4.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/NettyServerSingletons.java:23]: it is exposed already https://github.com/zeitlinger/opentelemetry-java-instrumentation/blob/6874a63723e7c1b1fc95535811a0a8f4114e4363/instrumentation/netty/netty-4.1/library/src/main/java/io/opentelemetry/instrumentation/netty/v4_1/NettyServerTelemetryBuilder.java#L46

just for javaagent, there is no dedicated property

File: instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpClientInstrumenterBuilder.java (hunk 7 above)
Comment (zeitlinger) [reply instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpClientInstrumenterBuilder.java:229]: maybe `toInstrumenterBuilder`?

File: instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpServerInstrumenterBuilder.java (hunk 8 above)
Comment (zeitlinger) [reply instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpServerInstrumenterBuilder.java:202]: no, they all do

Comment (zeitlinger) [reply instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpServerInstrumenterBuilder.java:202]: made it part of the ctor

**Comment by zeitlinger** [comment]
@trask please check again

File: instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpServerInstrumenterBuilder.java [hunk 9]
```diff
@@ -53,23 +52,26 @@ public final class DefaultHttpServerInstrumenterBuilder<REQUEST, RESPONSE> {
       httpAttributesExtractorBuilder;
   private final HttpSpanNameExtractorBuilder<REQUEST> httpSpanNameExtractorBuilder;
 
-  @Nullable private TextMapGetter<REQUEST> headerGetter;
+  private final TextMapGetter<REQUEST> headerGetter;
```
Comment (laurit) [inline instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpServerInstrumenterBuilder.java:55]: there is a null check for this field, so I guess it should have the `@Nullable`?

Comment (zeitlinger) [reply instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpServerInstrumenterBuilder.java:55]: it's actually never null as of now - but will be once ktor is converted to use the builder

Comment (zeitlinger) [reply instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpServerInstrumenterBuilder.java:55]: so I've added @Nullable back

File: instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpServerInstrumenterBuilder.java (hunk 8 above)
Comment (zeitlinger) [reply instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/builder/internal/DefaultHttpServerInstrumenterBuilder.java:197]: just found that ktor doesn't have the header getter - but this is an upcoming pr

**Comment by zeitlinger** [comment]
@laurit please check again

**Review by laurit — APPROVED** [review]

**Pull request merged**